### PR TITLE
v0.2

### PR DESCRIPTION
**Pull request by matklad**
This is a significant rewrite of xshell. The main change is the switch
from global-state API to object based API -- `Shell` is now a thing and
is the gateway to all API.

Motivation is manyfold:

* Curren `fs` API is unergonomic, as it requires importing many
  individual functions. With the new style API, `use xshell::{Shell,
  cmd};` is all you need, and API is discoverable via `sh.`
  auto-completion.
* Soundness. The recent discussion on IRLO
  <https://internals.rust-lang.org/t/synchronized-ffi-access-to-posix-environment-variable-functions/15475>
  convinced me that process' environment shold be read-only
* Implementation complexity: we spend a lot of effort debugging various
  issues related to `gsl`.

Specific changes:

* There's now a `Shell` which holds the environment. For ergonomics
  reasons, `Shell` uses interior mutability.
* The `cmd!` macro now requires a `Shell` as a first argument.
* The `Cmd` builder methods are somewhat more orthogonal.
* The suggested naming no longer follows UNIXism which makes the naming
  more incluiseve and natural (assuming you have code completion).

With the present API, I also feel we should shift from "convenient
hacks" to "robust tool". As a first step to that end, stdin is now fed
via a dedicated thread, to aviod deadlock.

This is *mostly* done I think, but I at least need to check this on windows as well.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -381,195 +769,267 @@ impl Cmd {
...
+    // endregion:builder
 
-    /// Returns the stdout from running the command.
-    pub fn read(self) -> Result<String> {
+    // region:running
+    /// Runs the command.
+    ///
+    /// By default the command itself is echoed to sterr, its standard streams
+    /// are inhereted, and non-zero return code is considered an error. These
+    /// behaviors be overriden by using various builder methods of the [`Cmd`].
```
Comment (azdavis) [inline src/lib.rs:869]:
```suggestion
    /// By default the command itself is echoed to stderr, its standard streams
    /// are inherited, and non-zero return code is considered an error. These
    /// behaviors can be overridden by using various builder methods of the [`Cmd`].
```

File: tests/compile_time.rs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+use xshell::{cmd, Shell};
+
+#[test]
+fn fixed_cost_compile_times() {
+    let sh = Shell::new().unwrap();
+
+    let _p = sh.push_dir("cbench");
+    let baseline = compile_bench(&sh, "baseline");
+    // FIXME: Don't have Interent rn, can't compile duct :-(
+    // let _ducted = compile_bench(&sh, "ducted");
```
Comment (azdavis) [inline tests/compile_time.rs:12]:
```suggestion
    // FIXME: Don't have internet rn, can't compile duct :-(
    // let _ducted = compile_bench(&sh, "ducted");
```

also, if you pushed this up, then you must have internet, so maybe you can rm this fixme? (:

File: src/lib.rs [hunk 3]
```diff
@@ -194,153 +229,500 @@
...
+    }
+
+    // region:env
+    /// Returns the working directory for this [`Shell`].
+    ///
+    /// All relative paths are interpreted relative to this directory, rather
+    /// than [`std::env::current_dir`].
+    pub fn current_dir(&self) -> PathBuf {
+        self.cwd.borrow().clone()
+    }
```
[the remark this answers is not in the document]
Comment (matklad) [reply src/lib.rs:319]: Sadly, this won't work due to `RefCell`. We *could* return a `Ref<Path>` here, but it seems too leaky.

File: src/lib.rs [hunk 4]
```diff
@@ -194,153 +229,500 @@
...
+        Cmd::new(self, program.as_ref())
+    }
+
+    fn path(&self, p: &Path) -> PathBuf {
+        let cd = self.cwd.borrow();
+        cd.join(p)
+    }
+}
+
+impl Shell {
```
[the remark this answers is not in the document]
Comment (matklad) [reply src/lib.rs:537]: Yeah, I am unsure myself -- the main value I see here is that someone with UNIX experience can just search the docs for `cd` and the like. Ie, I don't expect people to call these methods....

Aha! I thin what I need here is https://doc.rust-lang.org/rustdoc/advanced-features.html#add-aliases-for-an-item-in-documentation-search actualy

File: tests/compile_time.rs (hunk 2 above)
Comment (matklad) [reply tests/compile_time.rs:12]: Indeed!

**Comment by matklad** [comment]
Thanks for the review @azdavis!

Whats your overall feeling on the direction this is going in?

**Comment by ngirard** [comment]
(I'm just passing by, but I wanted to say: it's a damn good job guys! Keep going !)

**Comment by azdavis** [comment]
I like less global state, and renaming the fs functions to be more searchable is good too!

**Pull request merged**